### DISCUSSION TITLE
Cannot create signer: No installed provider supports this key: net.jsign.jca.SigningServicePrivateKey

### DISCUSSION THREAD

**Discussion by manishgant** [ebourg/jsign#188] [Q&A] [answered]
I am running into the above mentioned error _cannot create signer: No installed provider supports this key: net.jsign.jca.SigningServicePrivateKey_

What provider do I add to enable support for the SigningServicePrivateKey? I am currently using BouncyCastleProvider and I want to use AWS KMS with jsign to codesign inside a Java App.

**Comment by ebourg** [comment]
Could you post the code leading to this error?

**Reply by manishgant** [reply]
I have this in the Bean's initialization
```java
// Using Default AWS credentials obtained from instance
File pemFile = resourceLoader.getResource(pemFilePath);
FileUtils.copyInputStreamToFile(inputStream, pemFile);
Keystore keyStore = new KeyStoreBuilder().storetype(KeyStoreType.AWS).certfile(pemFile).keystore(System.getenv("AWS_REGION")).build();
``` 

Later I use this during the signing 

```java
fileToBeSigned = new File(context.getOutputfilePath());
AuthenticodeSigner signer = new AuthenticodeSigner(keyStore, "key-alias", "");
signer.withProgramURL("myApp.com")
	  .withTimestamping(true)
	  .withTimestampingAuthority(timestampingUrl);

try (Signable file = Signable.of(fileToBeSigned)) {
		   signer.sign(file);
}
``` 
This error comes up after signer.sign(file) is executed. If I debug internally I see the error comes from the line
```java
ContentSigner shaSigner = contentSignerBuilder.build(this.privateKey);
``` 
https://github.com/ebourg/jsign/blob/2bdef865dab8eed8dcd95ac92b71f1b3db23f618/jsign-core/src/main/java/net/jsign/AuthenticodeSigner.java#L413

And this is the stacktrace
```shell
org.bouncycastle.operator.OperatorCreationException: cannot create signer: No installed provider supports this key: net.jsign.jca.SigningServicePrivateKey
2023-12-06T03:59:30.006529172Z 	at org.bouncycastle.operator.jcajce.JcaContentSignerBuilder.build(Unknown Source)
2023-12-06T03:59:30.007152005Z 	at net.jsign.AuthenticodeSigner.createSignedDataGenerator(AuthenticodeSigner.java:413)
2023-12-06T03:59:30.007174130Z 	at net.jsign.AuthenticodeSigner.createSignedData(AuthenticodeSigner.java:372)
2023-12-06T03:59:30.007264672Z 	at net.jsign.AuthenticodeSigner.sign(AuthenticodeSigner.java:348)
2023-12-06T03:59:30.007272630Z  ......
```

**Reply by ebourg** [reply]
The storepass parameter isn't specified, is this running on an EC2 instance?

**Reply by manishgant** [reply, marked as answer]
Storepass isn't the issue, I am able to reach KMS and obtain what I need from there. 

Actually I figured it out, I had to explicitly set the Security provider as SigningServiceJcaProvider. It works now

**Reply by ebourg** [reply]
You are right, and this isn't properly documented. You should be able to just use the Provider of the KeyStore built instead of creating an instance of SigningServiceJcaProvider, something like this:

    Keystore keystore = new KeyStoreBuilder().storetype(AWS).certfile(pemFile).keystore(region).build();
    AuthenticodeSigner signer = new AuthenticodeSigner(keystore, "alias", null);
    signer.withProgramURL("myApp.com")
          .withSignatureProvider(keystore.getProvider())
          .withTimestamping(true);

**Reply by ebourg** [reply]
I've modified `AuthenticodeSigner` in 00f7a251 to automatically set the provider if the keystore is backed by a cloud signing service.
